### DISCUSSION TITLE
Single color environment for relfections

### DISCUSSION THREAD

**Discussion by IRobot1** [nartc/angular-three#106] [Ideas]
My understanding is that for a material to reflect, there has to be a environmental background.  I know that SOBA includes environment for setting the scene background using HDR image.  However, when background isn't visible, this causes objects to reflect something that isn't in the scene.

Does the SOBA environment support the ability to pass in my own HDR file?

Is there a possibility to create a fixed color environmental background for reflections?  The color would default to white, but could be tinted as needed.

**Comment by nartc** [comment]
@IRobot1 there will be updated to the Environment component in Soba and will most likely be similar to this https://drei.pmnd.rs/?path=/story/staging-environment--environment-story

**Reply by IRobot1** [reply]
Ok, so no support for non preset files.

The following code is failing and I can't figure out why.  
GET http://localhost:4200/assets/piza/p 404 (Not Found).  
GET http://localhost:4200/assets/piza/x 404 (Not Found).
GET http://localhost:4200/assets/piza/ 404 (Not Found)
GET http://localhost:4200/assets/piza/n 404 (Not Found)
GET http://localhost:4200/assets/piza/g 404 (Not Found)

It seems to be breaking the strings into individual characters and using that as the file name.

```ts
  constructor(private loader: NgtLoader) {
    const files= [
      'px.png',
      'nx.png',
      'py.png',
      'ny.png',
      'pz.png',
      'nz.png',
    ]
    this.loader.use(CubeTextureLoader, files, (innerLoader) => {
      innerLoader.setPath('assets/piza/');
    }).subscribe(next => {
    });
  }
```

**Reply by nartc** [reply]
You would want to do: `this.loader.use(CubeTextureLoader, [files]...)` instead.

**Comment by IRobot1** [comment]
files already is an array.

**Reply by nartc** [reply]
It still needs to be put in an array

**Comment by IRobot1** [comment]
![image](https://user-images.githubusercontent.com/25032599/165519435-c6d973ba-428f-4d22-b5a0-76c5f8ba03d7.png)

**Reply by nartc** [reply]
It is a bug with the typings of `NgtLoader`. You can force it with `as string[]`

**Comment by IRobot1** [comment]
That fixes the syntax error, but still no luck
![image](https://user-images.githubusercontent.com/25032599/165528012-b5a1d930-00b6-4dab-8017-3f50a5e6a40c.png)

Note that the following direct use of CubeTextureLoader works fine
```ts
    const files = ['px.png', 'nx.png', 'py.png', 'ny.png', 'pz.png', 'nz.png'];
    const ldrCubeMap = new CubeTextureLoader()
      .setPath('assets/white/')
      .load(files, function () {
        ldrCubeMap.encoding = sRGBEncoding;
      });
    this.texture = ldrCubeMap;
```

**Reply by nartc** [reply]
Weird. Because that's how the Environment component is doing.https://v5--angular-three.netlify.app/soba/?path=/story/staging-environment--files

**Comment by IRobot1** [comment]
> Is there a possibility to create a fixed color environmental background for reflections? The color would default to white, but could be tinted as needed. 

Using a preset background for reflection, you get this.

![image](https://user-images.githubusercontent.com/25032599/165548805-00ed8124-89ac-4f74-8dcd-36f7eccb48ac.png)

It is possible to have a white environment background.  However, without directional or spot light lights, the result is a flat color with a white border.

![image](https://user-images.githubusercontent.com/25032599/165547938-f43c10bb-8e86-4a01-a76c-69f1c2a29e74.png)

Here's the settings I used for the mesh
```html
<ngt-mesh [position]="[0,2,0]" castShadow>
  <ngt-sphere-geometry></ngt-sphere-geometry>
  <ngt-mesh-standard-material metalness="0.9" roughness="0.1"
                              color="violet"></ngt-mesh-standard-material>
</ngt-mesh>
```
However, after adding a point light, it looks better
```html
<ngt-point-light ngtPointLightHelper [position]="[2,3,2]" color="red"></ngt-point-light>
```
![image](https://user-images.githubusercontent.com/25032599/165547522-33acbd06-022a-48f3-97e8-8f6b76d1d3fe.png)

Here's the code to load and set the scene environment to a white background.
```ts
    const files = ['px.png', 'nx.png', 'py.png', 'ny.png', 'pz.png', 'nz.png'];
    const ldrCubeMap = new CubeTextureLoader()
      .setPath('assets/white/')
      .load(files, function () {
        ldrCubeMap.encoding = sRGBEncoding;
      });
    scene.environment = ldrCubeMap;
```
This this white image for all 6 files.
![nx](https://user-images.githubusercontent.com/25032599/165547263-d4f31802-506e-4aab-a77d-58f84b9641e9.png)

It should be possible to modify this to extract the image from a temporary 2d canvas filled with a color tint (or white).

Using this gray background removes the washed out edges
![nx](https://user-images.githubusercontent.com/25032599/165550753-b1be4dc0-2cac-4ee5-ace1-786ec9541d12.png)

![image](https://user-images.githubusercontent.com/25032599/165550815-6613e584-9eb5-4a1c-9133-0d7083d5ade9.png)
